### ISSUE TITLE
Java API: My java program abnormal exit while accessing Optimizer Check

### ISSUE THREAD

**Issue by jerry-tom** [Z3Prover/z3#6421]
My java program do optimizer check by z3，but sometimes the proram abnormal exit. 
Here are some excerpts from the error report file that I think might be related with the error:
``` text
tack: [0x000000030e4e8000,0x000000030e5e8000],  sp=0x000000030e5e5be0,  free space=1014k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [libz3.dylib+0xaa59dd]  model_evaluator::operator()(expr*)+0xd
C  [libz3.dylib+0xab4b02]  model::operator()(expr*)+0x12
C  [libz3.dylib+0xbc32f]  opt::opt_solver::maximize_objective(unsigned int, obj_ref<expr, ast_manager>&)::$_0::operator()() const+0x6f
C  [libz3.dylib+0xbbf44]  opt::opt_solver::maximize_objective(unsigned int, obj_ref<expr, ast_manager>&)+0x154
C  [libz3.dylib+0xc483f]  opt::optsmt::geometric_lex(unsigned int, bool)+0x3df
C  [libz3.dylib+0xc7f07]  opt::optsmt::lex(unsigned int, bool)+0x67
C  [libz3.dylib+0x8b826]  opt::context::execute_min_max(unsigned int, bool, bool, bool)+0x56
C  [libz3.dylib+0x8a00c]  opt::context::optimize(ref_vector<expr, ast_manager> const&)+0x61c
C  [libz3.dylib+0x4bf87]  Z3_optimize_check+0x1d7
C  [libz3java.dylib+0x82e5]  Java_com_microsoft_z3_Native_INTERNALoptimizeCheck+0x45
J 7059  com.microsoft.z3.Native.INTERNALoptimizeCheck(JJI[J)I (0 bytes) @ 0x0000000116192d15 [0x0000000116192c40+0xd5]
J 7058 C1 com.microsoft.z3.Native.optimizeCheck(JJI[J)I (48 bytes) @ 0x0000000116192504 [0x0000000116192480+0x84]
J 7057 C1 com.microsoft.z3.Optimize.Check([Lcom/microsoft/z3/Expr;)Lcom/microsoft/z3/Status; (96 bytes) @ 0x0000000116191934 [0x00000001161913c0+0x574]
J 6923 C1 dataflow.interval.IntervalAnalysis.getOptimizedLimitedValue(Ljava/util/List;Lcom/microsoft/z3/Context;Lcom/microsoft/z3/BoolExpr;Lcom/microsoft/z3/IntExpr;ZLdataflow/interval/IntervalLimitValue;)Ldataflow/interval/IntervalLimitValue; (312 bytes) @ 0x00000001161399dc [0x0000000116137a60+0x1f7c]
J 7086 C1 dataflow.interval.IntervalAnalysis.getOptimizedValue(Ldataflow/interval/IntervalValue;Ljava/util/List;Lcom/microsoft/z3/Context;Lcom/microsoft/z3/BoolExpr;Lcom/microsoft/z3/IntExpr;)Ldataflow/interval/IntervalValue; (123 bytes) @ 0x00000001161ba4bc [0x00000001161ba300+0x1bc]
J 7297 C1 dataflow.interval.IntervalAnalysis.handleWithSat(Lpath/IntervalPathAnalyzer;Ldataflow/interval/fact/IAFact;)V (218 bytes) @ 0x00000001162d6c7c [0x00000001162d4bc0+0x20bc]
```

Note:  the version of z3  is [z3-4.8.14](https://github.com/Z3Prover/z3/releases/tag/z3-4.8.14)

**Comment by NikolajBjorner** [comment]
it would be some bug unrelated to java. Can you reproduce it using the Log.java feature or dump the state of the optimize object as smt (the default toString method should do it) before the call to check and repro?

**Comment by jerry-tom** [comment]
ok, I can try to reproduce it, but it is episodic

here some additional information from the error report
``` text
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x000000013ae4d9dd, pid=11873, tid=0x0000000000014fc7
#
# JRE version: OpenJDK Runtime Environment (8.0_292-b10) (build 1.8.0_292-b10)
# Java VM: OpenJDK 64-Bit Server VM (25.292-b10 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# C  [libz3.dylib+0xaa59dd]  model_evaluator::operator()(expr*)+0xd
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# If you would like to submit a bug report, please visit:
#   https://github.com/AdoptOpenJDK/openjdk-support/issues
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#

---------------  T H R E A D  ---------------

Current thread (0x00007fdbf5281000):  JavaThread "Thread-1616" [_thread_in_native, id=85959, stack(0x000000030e4e8000,0x000000030e5e8000)]

siginfo: si_signo: 11 (SIGSEGV), si_code: 1 (SEGV_MAPERR), si_addr: 0x0000000000000080

Registers:
RAX=0x0000000000000000, RBX=0x000000030e5e5c68, RCX=0x00007fdbe5330050, RDX=0x00007fdbf5bee540
RSP=0x000000030e5e5be0, RBP=0x000000030e5e5bf0, RSI=0x0000000000000080, RDI=0x000000030e5e5c68
R8 =0x0000000000000003, R9 =0x0101010101010101, R10=0x0f0f0f0f0f0f0f0f, R11=0x3333333333333333
R12=0x0000000000000019, R13=0x00007fdbe54dd6f8, R14=0x0000000000000080, R15=0x00007fdbe54dda40
RIP=0x000000013ae4d9dd, EFLAGS=0x0000000000000202, ERR=0x0000000000000004
  TRAPNO=0x000000000000000e

Top of Stack: (sp=0x000000030e5e5be0)
0x000000030e5e5be0:   000000030e5e5c68 000000030e5e5e40
0x000000030e5e5bf0:   000000030e5e5c10 000000013ae5cb02
0x000000030e5e5c00:   0000000000000000 00007fdbe54dd6f8
0x000000030e5e5c10:   000000030e5e5d10 000000013a46432f
0x000000030e5e5c20:   0000000000000000 0000000000000001
0x000000030e5e5c30:   0000000000000000 00007f0000000000
0x000000030e5e5c40:   0000000000000000 0000000000000001
0x000000030e5e5c50:   0000000000000000 000000030e5e5d80
0x000000030e5e5c60:   0000000000000000 0000000000000000
0x000000030e5e5c70:   0000000000000001 0000000000000000
0x000000030e5e5c80:   0000000000000000 0000000000000001
0x000000030e5e5c90:   0000000000000000 0000000000000000
0x000000030e5e5ca0:   000000013b5d5290 00007fdbf609bcc8
0x000000030e5e5cb0:   0000000000000005 00007fdbe54dd6f8
0x000000030e5e5cc0:   00000019ee631a08 000000030e5e5d80
0x000000030e5e5cd0:   0000000000000019 00007fdbe54dd6f8
0x000000030e5e5ce0:   000000030e5e5d10 000000013a90d02f
0x000000030e5e5cf0:   000000013b5d5268 000000013b5d5268
0x000000030e5e5d00:   00007fdbe54dda68 00007fdbe54dda40
0x000000030e5e5d10:   000000030e5e5e80 000000013a463f44
0x000000030e5e5d20:   00007fdbee7bb150 00007fdbee7bb1e8
0x000000030e5e5d30:   00007fdbe691efa0 00007fdbe5475190
0x000000030e5e5d40:   00007fdbe691efa0 00007fdbee7bb008
0x000000030e5e5d50:   0000000000000000 000000030e5e5e30
0x000000030e5e5d60:   000000030e5e5e20 00007ff8014ec685
0x000000030e5e5d70:   00007fdbee7bb040 000000030e5e5dd8
0x000000030e5e5d80:   00007fd800000001 0000000000000000
0x000000030e5e5d90:   0000000000000001 0000000000000000
0x000000030e5e5da0:   00007fd800000000 0000000000000000
0x000000030e5e5db0:   00007fd800000001 0000000000000000
0x000000030e5e5dc0:   cfbc57cc00000000 0000000000000000
0x000000030e5e5dd0:   0000000000000001 0000000000000000 

Instructions: (pc=0x000000013ae4d9dd)
0x000000013ae4d9bd:   18 02 00 00 48 03 3b 4c 89 f6 5b 41 5e 5d e9 a0
0x000000013ae4d9cd:   0f 00 00 55 48 89 e5 41 56 53 49 89 f6 48 89 fb
0x000000013ae4d9dd:   48 8b 3e 48 8b 47 08 48 c7 03 00 00 00 00 48 89
0x000000013ae4d9ed:   43 08 48 8d 8f f0 01 00 00 48 89 d6 48 89 da e8 

Register to memory mapping:

RAX=0x0000000000000000 is an unknown value
RBX=0x000000030e5e5c68 is pointing into the stack for thread: 0x00007fdbf5281000
RCX=0x00007fdbe5330050 is an unknown value
RDX=0x00007fdbf5bee540 is an unknown value
RSP=0x000000030e5e5be0 is pointing into the stack for thread: 0x00007fdbf5281000
RBP=0x000000030e5e5bf0 is pointing into the stack for thread: 0x00007fdbf5281000
RSI=0x0000000000000080 is an unknown value
RDI=0x000000030e5e5c68 is pointing into the stack for thread: 0x00007fdbf5281000
R8 =0x0000000000000003 is an unknown value
R9 =0x0101010101010101 is an unknown value
R10=0x0f0f0f0f0f0f0f0f is an unknown value
R11=0x3333333333333333 is an unknown value
R12=0x0000000000000019 is an unknown value
R13=0x00007fdbe54dd6f8 is an unknown value
R14=0x0000000000000080 is an unknown value
R15=0x00007fdbe54dda40 is an unknown value
```

**Comment by jerry-tom** [comment]
we have some new stack info from the generated core file. The this pointer of model evaluator object is null in frame 7 and frame 8.

``` text
#0  0x00007fca6c673e7f in raise () from /lib64/libc.so.6
#1  0x00007fca6c65e8b5 in abort () from /lib64/libc.so.6
#2  0x00007fca6bf44835 in os::abort(bool) () from /usr/java/TencentKona-8.0.9-322/jre/lib/amd64/server/libjvm.so
#3  0x00007fca6c0f16d3 in VMError::report_and_die() () from /usr/java/TencentKona-8.0.9-322/jre/lib/amd64/server/libjvm.so
#4  0x00007fca6bf4b602 in JVM_handle_linux_signal () from /usr/java/TencentKona-8.0.9-322/jre/lib/amd64/server/libjvm.so
#5  0x00007fca6bf407b3 in signalHandler(int, siginfo*, void*) () from /usr/java/TencentKona-8.0.9-322/jre/lib/amd64/server/libjvm.so
#6  {<}signal handler called{>}
#7  0x00007fca1dac906c in model_evaluator::m (this=0x80) at /data/third-party/z3-z3-4.11.2/src/model/model_evaluator.cpp:753
#8  model_evaluator::operator() (this=0x80, t=0x7fc985c80988) at /data/third-party/z3-z3-4.11.2/src/model/model_evaluator.cpp:753
#9  0x00007fca1dabc6bd in model::operator() (this={<}optimized out{>}, t={<}optimized out{>}) at /data/third-party/z3-z3-4.11.2/src/model/model.cpp:551
#10 0x00007fca1e8933ed in opt::opt_solver::{<}lambda(){>}::operator() (__closure={<}synthetic pointer{>}, __closure={<}synthetic pointer{>})
    at /data/third-party/z3-z3-4.11.2/src/util/vector.h:478
#11 opt::opt_solver::{<}lambda(){>}::operator() (__closure={<}synthetic pointer{>}) at /data/third-party/z3-z3-4.11.2/src/opt/opt_solver.cpp:291
#12 opt::opt_solver::maximize_objective (this=0x7fc9781cfce8, i={<}optimized out{>}, i@entry=0, blocker=...) at /data/third-party/z3-z3-4.11.2/src/opt/opt_solver.cpp:307
#13 0x00007fca1e887a90 in opt::optsmt::geometric_lex (this=0x7fc9781cdba0, obj_index={<}optimized out{>}, is_maximize=true)
    at /data/third-party/z3-z3-4.11.2/src/opt/optsmt.cpp:224
#14 0x00007fca1e88d5b8 in opt::optsmt::lex (this=this@entry=0x7fc9781cdba0, obj_index=0, is_maximize=is_maximize@entry=true)
    at /data/third-party/z3-z3-4.11.2/src/opt/optsmt.cpp:509
#15 0x00007fca1e849e6b in opt::context::execute_min_max (this=0x7fc9781cda98, index=0, committed={<}optimized out{>}, scoped={<}optimized out{>}, is_max=true)
    at /data/third-party/z3-z3-4.11.2/src/opt/opt_context.cpp:443
#16 0x00007fca1e855623 in opt::context::optimize (this=0x7fc9781cda98, _asms=...) at /data/third-party/z3-z3-4.11.2/src/opt/opt_context.cpp:356
#17 0x00007fca1e8e8703 in Z3_optimize_check (c=c@entry=0x7fc985c85968, o=o@entry=0x7fc9781bc378, num_assumptions=num_assumptions@entry=0, 
    assumptions=assumptions@entry=0x7fca54002000) at /data/third-party/z3-z3-4.11.2/src/api/api_opt.cpp:163
#18 0x00007fca2407a274 in Java_com_microsoft_z3_Native_INTERNALoptimizeCheck (jenv=0x7fca649cb268, cls={<}optimized out{>}, a0=140503509653864, a1=140503280239480, a2=0, 
    a3=0x7fca1548c878) at /data/third-party/z3-z3-4.11.2/build_relwithdeginfo/src/api/java/Native.cpp:2923

```

**Comment by NikolajBjorner** [comment]
These stacks are unless without repros.
I suggest two ways to attempt a repro: Use the Log.java facility or dump a repro as SMTLIB2 benchmark.

**Comment by jerry-tom** [comment]
it happened by chance, and the core exception not triggerred by a fixed optimize object. I picked out several objects just printed before exception as follows:
1. (declare-fun a () Int)
(declare-fun b () Int)
(declare-fun c () Int)
(declare-fun d () Int)
(declare-fun e () Int)
(assert (and (< e (* b d))
     (< c (* b a))))
(assert (>= e 1))
(assert (>= b 1))
(minimize c)
(check-sat)

2. (declare-fun a () Int)
(declare-fun b () Int)
(declare-fun c () Int)
(assert (not (> c (* b a))))
(assert (>= c 0))
(assert (>= a 0))
(minimize b)
(check-sat)

it worked with single thead throughout the process. A global single instance of Context Class makes a new optimize object every time when calculating, and timeout limit set to 150 ms.

**Comment by NikolajBjorner** [comment]
they are non-linear and unbounded. Not really in scope of what is well-treated.

**Comment by jerry-tom** [comment]
Any other way to handle with the problem out of scope, rather than simply make the process break down? .e.g. give some response back to user identifying that it's a non-linear problem.

**Comment by jerry-tom** [comment]
Through dbg debugging analysis, we found that core is because m_last_model is a null pointer, as shown below

<img width="348" alt="WeChatWorkScreenshot_46efef56-e3fb-48d9-8003-bff1a756b7ab" src="https://user-images.githubusercontent.com/5044158/198565956-89cffc3e-5d75-4dc9-af8d-9cd40b44c8d7.png">

![WeChatWorkScreenshot_d6ec92c2-e44e-404a-9444-f1cf2934c56a](https://user-images.githubusercontent.com/5044158/198566847-17643a72-cc97-47fd-ac99-28d772a50996.png)
![WeChatWorkScreenshot_04862b10-8bce-4690-b998-619bd9d61b32](https://user-images.githubusercontent.com/5044158/198566880-083ad1ae-e862-4da4-bdca-3508846e9a92.png)


--------------
We have temporarily avoided core by modifying it. Is this the correct way?
<img width="373" alt="WeChatWorkScreenshot_55bc9b2f-54d7-4f3d-adee-6c619612bd4e" src="https://user-images.githubusercontent.com/5044158/198566372-a8275af2-f289-496d-b636-27e31c7b3122.png">

**Comment by NikolajBjorner** [comment]
I can't say why the model should be null, but it is presumably possible if a process got canceled.
I have updated the code to add the null checks.
